pydevices/HtsDevices/acq2106_WRPG.py

```diff
@@ -0,0 +1,193 @@
+            # for each channel:
+            for x in numpy.nditer(chan_t_states):
+                all_t_times_states.append(x) #Appends arrays made of one element,
+
+        # Choosing only the transition times:
+        all_t_times = all_t_times_states[0::2]
+
+        # Removing duplicates and then sorting in ascending manner:
+        t_times = []
+        for i in all_t_times:
+            if i not in t_times:
+                t_times.append(i)
+
+        # t_times contains the unique set of transitions times used in the experiment:
+        t_times = sorted(numpy.float64(t_times))
+
+        # initialize the state matrix
+        rows, cols = (len(t_times), nchan)
+        state = [[0 for i in range(cols)] for j in range(rows)]
+
+        # Building the state matrix. For each channel, we traverse all the transition times to find those who are 
+        # in the particular channel. 
+        # If the transition time is in the channel, we copied its state into the state[i][j] element. 
+        # If a transition time does not appear in that channel, we keep the previous state for, i.e. the state doesn't change.
+        for j in range(nchan):
+            chan_t_states = self.__getattr__('OUTPUT_%3.3d' % (j+1))
+
+            for i in range(len(t_times)):
+
+                if i == 0:
+                    state[i][j] = 0
+                else:
+                    state[i][j] = state[i-1][j]
+                    
+                    # chan_t_states its elements are pairs of [ttimes, state]. e.g [[0.0, 0],[1.0, 1],...]
+                    # chan_t_states[0] are all the first elements of those pairs, i.e the trans. times: 
+                    # e.g [[1D0], [2D0], [3D0], [4D0] ... ]
+                    # chan_t_states[1] are all the second elements of those pairs, i.e. the states: 
+                    # e.g [[0],[1],...]
+                    for t in range(len(chan_t_states[0])):
+                        #Check if the transition time is one of the times that belongs to this channel:
+                        if t_times[i] == chan_t_states[0][t][0]:
+                            state[i][j] = int(chan_t_states[1][t][0])
+
+
+        # Building the string of 1s and 0s for each transition time:
+        binrows = []
+        for row in state:
+            rowstr = [str(i) for i in numpy.flip(row)]  # flipping the bits so that chan 1 is in the far right position
+            binrows.append(''.join(rowstr))
+
+        # Converting the original units of the transtion times in seconds, to micro-seconts:
+        times_usecs = []
+        for elements in t_times:
+            times_usecs.append(int(elements * 1E6)) #in micro-seconds
+        # Building a pair between the t_times and bin states:
+        stl_list = zip(times_usecs, binrows)
+
+        # Creating the STL file with the path and file name given in self.stl_file.data()
+        fd, temp_path = mkstemp(prefix='acq2106-dio482-', suffix='.stl')
```

Comment:
We *really* should not be using a temp file for this, did you ever reach out to dtacq to get them to add a non-file version of this function to their library?

Reply:
👍

Reply:
No, I haven't yet. I was thinking of bring up this question, since I can see that they are doing something similar in their class for the _ACQ400_MR_BASE with the hardcoded STL file.